pkg/core/native/native_neo.go

```diff
@@ -461,23 +467,17 @@ func (n *NEO) balanceFromBytes(si *state.StorageItem) (*big.Int, error) {
 	return &acc.Balance, err
 }
 
-func (n *NEO) distributeGas(ic *interop.Context, h util.Uint160, acc *state.NEOBalance) error {
+func (n *NEO) distributeGas(ic *interop.Context, h util.Uint160, acc *state.NEOBalance) (*big.Int, error) {
 	if ic.Block == nil || ic.Block.Index == 0 || ic.Block.Index == acc.BalanceHeight {
-		return nil
+		return nil, nil
 	}
 	gen, err := n.calculateBonus(ic.DAO, acc.VoteTo, &acc.Balance, acc.BalanceHeight, ic.Block.Index)
 	if err != nil {
-		return err
+		return nil, err
 	}
 	acc.BalanceHeight = ic.Block.Index
 
-	// Must store acc before GAS distribution to fix acc's BalanceHeight value in the storage for
-	// further acc's queries from `onNEP17Payment` if so, see https://github.com/nspcc-dev/neo-go/pull/2181.
-	key := makeAccountKey(h)
-	ic.DAO.PutStorageItem(n.ID, key, acc.Bytes())
```

Comment:
Why did you remove these lines? It's always `GetAndChange` in C# and two lines above we've changed acc's BalanceHeight. We must save account anyway here.

Reply:
It's just no longer needed, it'll be saved before we're to invoke `Mint` in `(c *nep17TokenNative) updateAccBalance()` or `(c *nep17TokenNative) addTokens()`.

Reply:
I still have doubds about this removal. It was here not only to save BalanceHeight, but also for the following case: if the subsequent code will return an error (e.g. `ModifyAccountVotes` inside `VoteInternal` will return an error), then no account will be saved in neo-go implementation (against the C# node where account's height is updated). NEO's `vote` doesn't panic if an error is returned from `VoteInternal`, instead it puts `false` on stack. Here comes the state diff.

Reply:
Although there's a *very* tiny chance that `ModifyAccountVotes` will return an error.

Reply:
In this case the transfer will fail for sure and GAS shouldn't be minted, thus account height shouldn't be updated and DB can be left as is. Although for `VoteInternal` it can be solved.

Reply:
Wow. A simple
```
--- a/pkg/core/native/native_neo.go
+++ b/pkg/core/native/native_neo.go
@@ -812,6 +812,7 @@ func (n *NEO) VoteInternal(ic *interop.Context, h util.Uint160, pub *keys.Public
        if err != nil {
                return err
        }
+       ic.DAO.PutStorageItem(n.ID, key, acc.Bytes()) // It won't be changed after that, but the DB must be updated.
        if err := n.ModifyAccountVotes(acc, ic.DAO, new(big.Int).Neg(&acc.Balance), false); err != nil {
                return err
        }
@@ -819,7 +820,6 @@ func (n *NEO) VoteInternal(ic *interop.Context, h util.Uint160, pub *keys.Public
        if err := n.ModifyAccountVotes(acc, ic.DAO, &acc.Balance, true); err != nil {
                return err
        }
-       ic.DAO.PutStorageItem(n.ID, key, acc.Bytes())
        if newGas != nil { // Can be if it was already distributed in the same block.
                n.GAS.mint(ic, h, newGas, true)
        }
```
Breaks some tests.

Reply:
OK, NM, that's intended.

Reply:
> NEO's vote doesn't panic if an error is returned from VoteInternal,

I think where we have an error returned C# node will just `throw` and that will kill the execution.

Reply:
You're right, then we can safely remove it.